addons/lunch/models/lunch_order.py

```diff
@@ -41,10 +41,128 @@ class LunchOrder(models.Model):
+            # since in the interface you usually don't update more than one order at a time but this is a bug nonetheless
+            topping_1 = values.pop('topping_ids_1')[0][2] if 'topping_ids_1' in values else self[:1].topping_ids_1.ids
+            topping_2 = values.pop('topping_ids_2')[0][2] if 'topping_ids_2' in values else self[:1].topping_ids_2.ids
+            topping_3 = values.pop('topping_ids_3')[0][2] if 'topping_ids_3' in values else self[:1].topping_ids_3.ids
+        else:
+            topping_1 = values['topping_ids_1'][0][2] if 'topping_ids_1' in values else []
+            topping_2 = values['topping_ids_2'][0][2] if 'topping_ids_2' in values else []
+            topping_3 = values['topping_ids_3'][0][2] if 'topping_ids_3' in values else []
+
+        return topping_1 + topping_2 + topping_3
+
+    @api.constrains('topping_ids_1', 'topping_ids_2', 'topping_ids_3')
+    def _check_topping_quantity(self):
+        errors = {
+            '1_more': _('You should order at least one %s'),
+            '1': _('You have to order one and only one %s'),
+        }
+        for line in self:
+            for index in range(1, 4):
+                availability = line['available_toppings_%s' % index]
+                quantity = line['topping_quantity_%s' % index]
+                toppings = line['topping_ids_%s' % index].filtered(lambda x: x.topping_category == index)
+                label = line['topping_label_%s' % index]
+
+                if availability and quantity != '0_more':
+                    check = bool(len(toppings) == 1 if quantity == '1' else toppings)
+                    if not check:
+                        raise ValidationError(errors[quantity] % label)
+
+    @api.model
+    def create(self, values):
+        lines = self._find_matching_lines({
+            **values,
+            'toppings': self._extract_toppings(values),
+        })
+        if lines:
+            # FIXME This will update multiple lines in the case there are multiple matching lines
+            # which should not happen trough the interface
+            lines.update_quantity(1)
+            return lines[:1]
+        return super().create(values)
+
+    def write(self, values):
+        merge_needed = 'note' in values or 'topping_ids_1' in values or 'topping_ids_2' in values or 'topping_ids_3' in values
+
+        # Only write on topping_ids_1 because they all share the same table and we don't want to remove all the records
+        # _extract_toppings will pop topping_ids_1, topping_ids_2 and topping_ids_3 from values
+        # This also forces us to invalidate the cache for topping_ids_2 and topping_ids_3 that could have changed
+        # through topping_ids_1 without the cache knowing about it
+        toppings = self._extract_toppings(values)
+        values['topping_ids_1'] = [(6, 0, toppings)]
+        self.invalidate_cache(['topping_ids_2', 'topping_ids_3'])
```

Comment:
We could try to remove this `invalidate_cache` by converting the 6 commands to 4 commands for each fields and adding a 5 command on topping_ids_1 before all other commands are executed
But that could break if for any reasons topping_ids_1 is not the first field to be written to

eg: 
```python
self.write({
    'topping_ids_1': [(6, 0, (1, 2))], 
    'topping_ids_2': [(6, 0, (5,))],
})

# Would become
self.write({
    'topping_ids_1': [(5, 0, 0), (4, 0, (1, 2))],
    'topping_ids_2': [(4, 0, (5,))],
})

# But if topping_ids_2 is written before topping_ids_1 then it will break
# because the ORM will add id 5 and then remove everything from the table
# and then add ids 1 and 2
```